### PR TITLE
Toml support

### PR DESCRIPTION
**Pull request by zekth**
As told in https://github.com/denoland/deno_std/issues/232 here is a draft of the TOML parser.

Added the documentation for the module.
Also the handling of `Array Of Tables` is not finished yet.

**Questions**:
- If a declaration is invalid do we skip it or we throw an Exception? Also an option `strict` could be added to handle both cases.

**Note**: @GrosSacASac `deepAssign` function has been added to `util/deep_assign.ts`

[TOML specs](https://github.com/toml-lang/toml)

### CODE REVIEW

File: toml/parser.ts [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+};
+
+const regGroup = /\[.*\]/;
+class KeyValuePair {
+  key: string;
+  value: unknown;
+}
+class Parser {
+  tomlLines: string[];
+  context: any;
```
Comment (ry) [inline toml/parser.ts]: Can this be given a non-any type? (It seems lint shouldn't allow any)

File: toml/parser_test.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import { assertEquals } from "../testing/asserts.ts";
+import { parseFile } from "./parser.ts";
+import * as path from "../fs/path/mod.ts";
+const testFilesDir = path.resolve("toml", "test");
+test(function tomlParseCRLF() {
+  const expected = { boolean: { bool1: true, bool2: false } };
+  const actual = parseFile(path.join(testFilesDir, "CRLF.toml"));
+  assertEquals(actual, expected);
+});
+test(function tomlParseboolean() {
```
Comment (ry) [inline toml/parser_test.ts]: line breaks between test cases please.

File: toml/parser.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+import { existsSync } from "../fs/exists.ts";
+
+const deepAssign = (target: object, ...sources: object[]): object => {
```
Comment (ry) [inline toml/parser.ts]: Can you move this to `//util/deep_assign.ts`

Also https://deno.land/style_guide.html#toplevelfunctionsshouldnotusearrowsyntax

Comment (zekth) [reply toml/parser.ts]: was a draft, tried to make it work. Also will add tests for it as we will expose it direclty in std.

File: toml/parser.ts (hunk 1 above)
Comment (zekth) [reply toml/parser.ts]: Have to do a bit of refactor to make it work but it's in the scope yep.

**Comment by zekth** [comment]
Does someone have opinion about questions?

cc @hayd @ry @j-f1 ?

File: toml/parser.ts (hunk 1 above)
Comment (zekth) [reply toml/parser.ts]: corrected

**Comment by bartlomieju** [comment]
@zekth re: `strict` option, AFAIK TOML has no strict/non-strict mode, so IMHO it should throw exception when declaration is invalid.

**Comment by ry** [comment]
@zekth are there any existing libraries we can examine to copy behavior from?

**Comment by zekth** [comment]
> @zekth are there any existing libraries we can examine to copy behavior from?

There are some in the [TOML wiki](https://github.com/toml-lang/toml/wiki), like:
- https://www.npmjs.com/package/@iarna/toml
- https://www.npmjs.com/package/ion-parser

**Comment by ry** [comment]
@zekth And how do they act on invalid declaration?

**Comment by zekth** [comment]
@ry directly throws exception. It was just an idea i had.

File: toml/README.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# TOML
+
+This module parse TOML files. It follows as much as possible the [TOML specs](https://github.com/toml-lang/toml). Be sure to read the supported types as not every specs is supported at the moment and the handling in TypeScript side is a bit different.
```
Comment (ry) [inline toml/README.md]: Lines should be wrapped at 80 columns. Please run the "format.ts" program on this README

File: toml/parser.ts [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+  key: string;
+  value: unknown;
+}
+class ParserGroup {
+  type: string;
+  name: string;
+  arrValues: unknown[] = [];
+  objValues: object = {};
+}
+class ParserContext {
```
Comment (ry) [inline toml/parser.ts]: Please add a line break between these top level declarations

File: toml/test/arrays.toml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+[arrays]
```
Comment (ry) [inline toml/test/arrays.toml]: Please rename the toml/test directory to toml/testdata

File: util/deep_assign.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+export function deepAssign(target: object, ...sources: object[]): object {
```
Comment (ry) [inline util/deep_assign.ts]: Thanks for moving this.

**Comment by zekth** [comment]
@ry you put it as ready for review, there is still a remaining type handling left. I'm working on it and on some speed optimisations. Do you want to merge it in this state or wait for the last type to be handled? No problem for me in both cases.

**Comment by ry** [comment]
@zekth I'm ok with merging now - looks useful as it is - but please complete that in follow up PRs

**Comment by zekth** [comment]
@ry ready to ship. Array of tables is handled now, also updated the doc for the warning. I'll do some optimisation in future PRs to fix the speed and remaining cases.

**Comment by ry** [comment]
Can you copy https://github.com/denoland/deno/blob/master/Cargo.toml to `toml/testdata/deno_cargo.toml` and check that this can parse it?

**Comment by zekth** [comment]
@ry Ready to merge.

**Pull request merged**